chia/wallet/trading/offer.py

```diff
@@ -0,0 +1,385 @@
+from dataclasses import dataclass
+from typing import List, Optional, Dict, Set, Tuple
+from blspy import G2Element
+
+from chia.types.blockchain_format.sized_bytes import bytes32
+from chia.types.blockchain_format.coin import Coin
+from chia.types.blockchain_format.program import Program
+from chia.types.announcement import Announcement
+from chia.types.coin_spend import CoinSpend
+from chia.types.spend_bundle import SpendBundle
+from chia.util.ints import uint64
+from chia.wallet.cc_wallet.cc_utils import (
+    CC_MOD,
+    SpendableCC,
+    construct_cc_puzzle,
+    match_cat_puzzle,
+    unsigned_spend_bundle_for_spendable_ccs,
+)
+from chia.wallet.lineage_proof import LineageProof
+from chia.wallet.puzzles.load_clvm import load_clvm
+from chia.wallet.payment import Payment
+
+OFFER_MOD = load_clvm("settlement_payments.clvm")
+ZERO_32 = bytes32([0] * 32)
+
+
+@dataclass(frozen=True)
+class NotarizedPayment(Payment):
+    nonce: bytes32 = ZERO_32
+
+    @classmethod
+    def from_condition_and_nonce(cls, condition: Program, nonce: bytes32) -> "NotarizedPayment":
+        with_opcode: Program = Program.to((51, condition))  # Gotta do this because the super class is expecting it
+        p = Payment.from_condition(with_opcode)
+        puzzle_hash, amount, memos = tuple(p.as_condition_args())
+        return cls(puzzle_hash, amount, memos, nonce)
+
+
+@dataclass(frozen=True)
+class Offer:
+    requested_payments: Dict[bytes32, List[NotarizedPayment]]  # The key is the asset id of the asset being requested
+    bundle: SpendBundle
+
+    @staticmethod
+    def ph():
+        return OFFER_MOD.get_tree_hash()
+
+    @staticmethod
+    def notarize_payments(
+        requested_payments: Dict[Optional[bytes32], List[Payment]],  # `None` means you are requesting XCH
+        coins: List[Coin],
+    ) -> Dict[Optional[bytes32], List[NotarizedPayment]]:
+        # This sort should be reproducible in CLVM with `>s`
+        sorted_coins: List[Coin] = sorted(coins, key=Coin.name)
+        sorted_coin_list: List[List] = [c.as_list() for c in sorted_coins]
+        nonce: bytes32 = Program.to(sorted_coin_list).get_tree_hash()
```

Comment:
This is probably a stretch, and I'm not sure what the outcome would be, but:

If I create an offer that includes a set of coins, and then cancel that offer insecurely... Creating a new offer with that same set of coins should yield the same nonce, right? What if that original offer was actually shared and taken by another party?

Reply:
The nonce's purpose is to prevent two offers from being spent together that assert the same set of payments.  By calculating the nonce based on the coins that are spent, we ensure that the only way to get a similar nonce is to offer the exact same coins.  Therefore there is no way for you to push both offers because the coins can only be spent one of the two ways.